Fixing race condition in yesod-bin

psibi opened the pull request:
Stack build process emits line even after successful build process
which leads to the overwriting of the appPortVar with -1. This leads
it to a compile mode again. Pressing Return Key and rebuilding it
again makes it go, but that's just a workaround I have to do every now
and then to solve the actual issue.

I'm using a `MVar` based locking solution for fixing the race
condition introduced.

I have been using this patch for around 3 to 4 days in our workplace and we haven't seen any surprising behavior so far.

snoyberg commented on yesod-bin/Devel.hs:
```diff
@@ -222,6 +225,33 @@ checkDevelFile =
+
+makeEmptyMVar :: MVar a -> IO ()
+makeEmptyMVar mvar = do
+  isEmpty <- isEmptyMVar mvar
```
It's safer to use tryTakeMVar

snoyberg commented on yesod-bin/Devel.hs:
```diff
@@ -222,6 +225,33 @@ checkDevelFile =
+
+updateAppPort :: ByteString -> MVar (BuildOutput) -> TVar Int -> IO ()
+updateAppPort bs mvar appPortVar = do
+  isEmpty <- isEmptyMVar mvar
```
It seems like the logic here would be easier to follow as a TVar Bool instead of MVar Build output.

psibi commented:
@snoyberg Thanks for the PR, I have updated the PR. I have also added `DEBUG` flag to cabal file which IMO helps during development of `yesod-bin`. I will remove it if you feel it's not needed.

snoyberg commented on yesod-bin/Devel.hs:
```diff
@@ -222,6 +227,42 @@ checkDevelFile =
+                                         -- yet.
+              -> TVar Int -> IO ()
+updateAppPort bs buildStarted appPortVar = do
+  hasStarted <- readTVarIO buildStarted
```
This looks like a race condition to me, at least in principle. Why not turn the entire function into a single `STM` block and perform the read and modifications atomically?

snoyberg commented:
I'm not opposed to some debugging output, but I'd rather see it happen with a `--verbose` flag than CPP. I'm indifferent on whether we keep the CPP in for now though.

psibi commented:
Thanks for the review. Indeed making it as a single STM block is better. I have updated the PR.

snoyberg commented on yesod/yesod.cabal:
```diff
@@ -1,5 +1,5 @@
 name:            yesod
-version:         1.4.5
+version:         1.4.6
```
Not sure if this was here before or not, is it supposed to be part of this PR?

psibi replied:
Sorry, that was unintended. Thanks for the catch, have reverted this.

snoyberg commented:
LGTM, thanks for the update. Merge when ready 👍

psibi commented:
Thanks, will merge it once that Travis CI passes.

psibi commented:
I would love to see a new release of this version in Hackage.
